cli.go

```diff
@@ -268,10 +342,109 @@ func (c *CLI) init() {
 		}
 	}
 
+	// Setup autocomplete if we have it enabled. We have to do this after
+	// the command tree is setup so we can use the radix tree to easily find
+	// all subcommands.
+	if c.Autocomplete {
+		c.initAutocomplete()
+	}
+
 	// Process the args
 	c.processArgs()
 }
 
+func (c *CLI) initAutocomplete() {
+	if c.AutocompleteInstall == "" {
+		c.AutocompleteInstall = defaultAutocompleteInstall
+	}
+
+	if c.AutocompleteUninstall == "" {
+		c.AutocompleteUninstall = defaultAutocompleteUninstall
+	}
+
+	if c.autocompleteInstaller == nil {
+		c.autocompleteInstaller = &realAutocompleteInstaller{}
+	}
+
+	// Build the root command
+	cmd := c.initAutocompleteSub("")
+
+	// For the root, we add the global flags
+	cmd.GlobalFlags = map[string]complete.Predictor{
+		"-" + c.AutocompleteInstall:   complete.PredictNothing,
+		"-" + c.AutocompleteUninstall: complete.PredictNothing,
+		"-help":    complete.PredictNothing,
+		"-version": complete.PredictNothing,
+	}
+	for k, v := range c.AutocompleteGlobalFlags {
+		cmd.GlobalFlags[k] = v
+	}
+
+	c.autocomplete = complete.New(c.Name, cmd)
+}
+
+func (c *CLI) initAutocompleteSub(prefix string) complete.Command {
+	var cmd complete.Command
+	walkFn := func(k string, raw interface{}) bool {
+		if len(prefix) > 0 {
+			// If we have a prefix, trim the prefix + 1 (for the space)
+			// Example: turns "sub one" to "one" with prefix "sub"
+			k = k[len(prefix)+1:]
+		}
+
+		// Keep track of the full key so that we can nest further if necessary
+		fullKey := k
+
+		if idx := strings.LastIndex(k, " "); idx >= 0 {
+			// If there is a space, we trim up to the space
+			k = k[:idx]
+		}
+
+		if idx := strings.LastIndex(k, " "); idx >= 0 {
+			// This catches the scenario just in case where we see "sub one"
+			// before "sub". This will let us properly setup the subcommand
+			// regardless.
+			k = k[idx+1:]
+		}
+
+		if _, ok := cmd.Sub[k]; ok {
+			// If we already tracked this subcommand then ignore
+			return false
+		}
+
+		if cmd.Sub == nil {
+			cmd.Sub = complete.Commands(make(map[string]complete.Command))
+		}
+		subCmd := c.initAutocompleteSub(fullKey)
+
+		// Instantiate the command so that we can check if the command is
+		// a CommandAutocomplete implementation. If there is an error
+		// creating the command, we just ignore it since that will be caught
+		// later.
+		impl, err := raw.(CommandFactory)()
+		if err != nil {
+			impl = nil
+		}
+
+		// Check if it implements ComandAutocomplete. If so, setup the autocomplete
+		if c, ok := impl.(CommandAutocomplete); ok {
```

Comment:
Would it be worth having these as two interfaces so they can be implemented separately?